tests/lib/services/gulag-runner.test.js

```diff
@@ -0,0 +1,212 @@
+    this.runner.countdownHandle = 'dummyhandle';
+    this.runner.addPrisoner('prisoner1');
+    this.runner.addPrisoner('prisoner2');
+    this.runner.matchDurationMilliseconds = 700;
+    this.runner.matchCountdownMilliseconds = 0;
+    this.runner.startMatchCountdown();
+    setTimeout(() => {
+      this.runner.currentMatch['prisoner1'] = [this.runner.currentMatch.answer, 123];
+      setTimeout(() => {
+        assert.deepStrictEqual(this.mockServices.punishmentStream.punishments, [
+          makeMute('prisoner2', this.runner.muteDurationSeconds, 'Lost in the gulag')
+        ]);
+        assert.deepStrictEqual(this.mockServices.messageRelay.messages.slice(0, 1), [
+          'prisoner2 and prisoner1 are now engaged in mortal combat in the gulag!'
+        ]);
+        assert.deepStrictEqual(this.mockServices.messageRelay.messages.slice(2), [
+          `They have ${this.runner.matchDurationMilliseconds / 1000} seconds to answer with !answer.`,
+          'The audience can use !stone <user> to cancel out a submitted answer',
+          'prisoner1 beat prisoner2 and won their freedom, prisoner2 did not answer correctly in time'
+        ]);
+        done()
+      }, 700);
+    }, 100)
+  });
+
+  it('can handle running a gulag match with a winner where both players answer correctly', (done) => {
+    this.runner.countdownHandle = 'dummyhandle';
+    this.runner.addPrisoner('prisoner1');
+    this.runner.addPrisoner('prisoner2');
+    this.runner.matchDurationMilliseconds = 700;
+    this.runner.matchCountdownMilliseconds = 0;
+    this.runner.startMatchCountdown();
+    setTimeout(() => {
+      this.runner.currentMatch['prisoner1'] = [this.runner.currentMatch.answer, 1];
+      this.runner.currentMatch['prisoner2'] = [this.runner.currentMatch.answer, 2];
+      setTimeout(() => {
+        assert.deepStrictEqual(this.mockServices.punishmentStream.punishments, [
+          makeMute('prisoner2', this.runner.muteDurationSeconds, 'Lost in the gulag')
+        ]);
+        assert.deepStrictEqual(this.mockServices.messageRelay.messages.slice(0, 1), [
+          'prisoner2 and prisoner1 are now engaged in mortal combat in the gulag!'
+        ]);
+        assert.deepStrictEqual(this.mockServices.messageRelay.messages.slice(2), [
+          `They have ${this.runner.matchDurationMilliseconds / 1000} seconds to answer with !answer.`,
+          'The audience can use !stone <user> to cancel out a submitted answer',
+          'prisoner1 beat prisoner2 by 0.001s and won their freedom'
+        ]);
+        done()
+      }, 700);
+    }, 100)
+  });
+
+  it('can handle running a gulag match with no winner where both players answer incorrectly', (done) => {
+    this.runner.countdownHandle = 'dummyhandle';
+    this.runner.addPrisoner('prisoner1');
+    this.runner.addPrisoner('prisoner2');
+    this.runner.matchDurationMilliseconds = 700;
+    this.runner.matchCountdownMilliseconds = 0;
+    this.runner.startMatchCountdown();
+    setTimeout(() => {
```

Comment:
https://github.com/destinygg/chat-bot/blob/master/tests/lib/services/punishment-cache.test.js


Don't use timeouts for tests, instead mock the clock using sinon, and control the advancement of it. The test I linked has an example of his this is done.

Reply:
Sure will do